src/query/api/v1/middleware/query_inspection.go

```diff
@@ -0,0 +1,230 @@
+// Copyright (c) 2021 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package middleware
+
+import (
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/prometheus/prometheus/promql/parser"
+	"github.com/uber-go/tally"
+
+	"github.com/m3db/m3/src/cmd/services/m3query/config"
+	"github.com/m3db/m3/src/query/api/v1/handler/prometheus/native"
+	"github.com/m3db/m3/src/x/headers"
+)
+
+const (
+	querySizeMetricName      = "size"
+	countThresholdMetricName = "count_threshold"
+	rangeThresholdMetricName = "range_threshold"
+	querySizeSmall           = "small"
+	querySizeLarge           = "large"
+
+	// promDefaultLookback is the default lookback in Prometheus, that affects
+	// the actual range it fetches datapoints from.
+	promDefaultLookback = time.Minute * 5
+)
+
+type queryInspectionMetrics struct {
+	notQuery            tally.Counter
+	notInspected        tally.Counter
+	belowCountThreshold tally.Counter
+	badQuery            tally.Counter
+	belowRangeThreshold tally.Counter
+	largeQuery          tally.Counter
+}
+
+func newQueryInspectionMetrics(scope tally.Scope) queryInspectionMetrics {
+	buildCounter := func(status string) tally.Counter {
+		return scope.Tagged(map[string]string{"status": status}).Counter("count")
+	}
+
+	return queryInspectionMetrics{
+		notQuery:            buildCounter("not_query"),
+		notInspected:        buildCounter("not_inspected"),
+		belowCountThreshold: buildCounter("below_count_threshold"),
+		badQuery:            buildCounter("bad_query"),
+		belowRangeThreshold: buildCounter("below_range_threshold"),
+		largeQuery:          buildCounter("large_query"),
+	}
+}
+
+func retrieveQueryRange(expr parser.Node, rangeSoFar time.Duration) time.Duration {
```

Comment:
worth adding unit tests on this logic?

Reply:
Yeah, may do; there's some coverage through the metrics_test providing a couple of different ranges, but may as well add some explicit ones around here since it's decently complex :+1: